src/strong_ptr.h

```diff
@@ -0,0 +1,317 @@
+// Copyright (c) 2017 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_STRONGPTR_H
+#define BITCOIN_STRONGPTR_H
+
+#include <memory>
+
+/*
+    strong_ptr and decay_ptr create a safe use pattern for moving shared_ptrs
+    to a unique_ptr.
+
+    While any such pattern would usually make no sense, it can be made to work
+    with a new type that holds an extra reference (strong_ptr), and one that
+    observes the lifetime of extra references (decay_ptr).
+
+    The process is:
+        1. Create a strong_ptr like any typical smart pointer. It cannot be
+            copied, but it can "loan out" shared_ptrs using
+            strong_ptr::get_shared().
+        2. Use the "loaned" shared_ptrs as usual, and use the strong_ptr as
+            though it were a unique_ptr.
+        3. When it's time to coalesce, move the strong_ptr into a decay_ptr.
+        4. Query the status with decay_ptr::decayed().
+
+    Once moved into a decay_ptr, the original strong_ptr is reset, and the
+    decay_ptr is unable to loan out any new shared_ptrs. Once the decay_ptr has
+    decayed, it behaves just like a unique_ptr.
+
+    If a loaned shared_ptr outlives the strong_ptr that owns it, the lifetime
+    is extended until the last copy is deleted. This is accomplished by
+    creating a shared_ptr to represent the lifetime of the strong_ptr, and
+    storing a copy of it in the loaned shared_ptr's deleter.
+
+    Thus, strong_ptr and decay_ptr ensure that allocated memory is always freed.
+
+    Here's a quick example of their use:
+
+        void func(std::shared_ptr<int>)
+        {
+            // Some long-lived operation
+        }
+
+        int main()
+        {
+            strong_ptr<int> str(make_strong<int>(100));
+            std::thread(func, str.get_shared());
+            decay_ptr<int> dec(std::move(strong));
+            // The original strong_ptr is now reset
+            while (!dec.decayed()) {
```

Comment:
I don't understand how make_strong/strong_ptr/decay_ptr provide any benefit in this example. Why wouldn't you just use regular shared pointers, and write this loop as `while (str.use_count() > 1)`?

Reply:
a weak_ptr can be created from a shared_ptr without bumping its refcount. That weak_ptr can ```lock()``` in a separate thread just after checking use_count() here. ```shared_ptr.unique()``` (```shared_ptr.use_count() == 1```) was deprecated in c++17 for that reason.

Once moved to a decay_ptr, ```decayed()``` is a trustworthy ```unique()```.

Reply:
I don't think this is a particularly good example, as it encourages busy waiting :-) But for the application in our net code it's a clever solution.